### DISCUSSION TITLE
Periodic visits to the same customer with buffer

### DISCUSSION THREAD

**Discussion by do-me** [reinterpretcat/vrp#118] [Q&A] [answered]
I was just wondering about a complex scenario and the best/most convenient way to model it (theoretically) in VRP. 

**Jobs**: many shops all over Germany; all shops require periodic service visits, some 1x/year, some 2x, 4x, 6x, 12x etc. which translates approximately to: 
- anytime during the year 
- once per 6 monthts, but there should be a minimum buffer of N months in between the 2 visits
- once per 3 monthts, but there should be a minimum buffer of N months in between the 4 visits etc.

additional constraint: one shop must always be visited from the same person (easy with VRP's `groups`)

**Fleet**: many service providers with different starting points; e.g. each 220 actual working days per year

In the past I was modeling it with a simple Python script more or less like this: 
- random starting offset for each shop to visit (e.g. +X working days)
- calculate the allowed range, e.g. for 2x per year, working days 0-80 are fine and then 140-220 so we'd have a guarantee of 60 working days buffer (this can be modeled with skills)
- repeat for all jobs and hope that there are no (random) extreme peaks for some periods so that some shops remain unserved

This was working ok-ish but it's a lot of trial and error work to play with parameters, such as offsets, different buffer values etc. to figure out a solution without peaks. 

I was searching through the VRP docs and found the [Periodic VRP (PVRP)](https://reinterpretcat.github.io/vrp/getting-started/features.html?highlight=periodic#features) which is sort of what I'd need.

The [compatibility](https://reinterpretcat.github.io/vrp/concepts/pragmatic/problem/jobs.html) param is already great for assigning the jobs to different days but afaik there is no way to avoid all let's say 12 yearly jobs being served consecutively in 12 working days.

Do you have better idea how to model this? 

P.S. We're in the phase of testing VRP for different complex scenarios and will share our results (probably after the summer).

**Comment by reinterpretcat** [comment, marked as answer]
>  [Periodic VRP (PVRP)](https://reinterpretcat.github.io/vrp/getting-started/features.html?highlight=periodic#features) which is sort of what I'd need.

The current implementation focuses on multiple vehicle shifts, not recurrent jobs. At the moment, the only one way to model such time buffer between serving logically recurrent jobs is their time windows. To satisfy your requirement natively, it would require to write a new constraint which should consider this minimal buffer. I think it is possible, but would require some careful implementation of time buffer handling, e.g. pushing already assigned jobs to later time when a new job is inserted (+ adding extra parameter to model it in pragmatic format)

**Reply by do-me** [reply]
Thanks for your answer. Actually, in the past I found it easier not to use precise dates or timestamps for the vehicles or jobs. I'd just refer to working days 1-220 e.g. as skills instead. In this way, it's a little more flexible considering holidays (and less confusing for the drivers when there is no calculated date). So I'll keep my Python script for now.
